src/TeamSpeak3QueryApi/Specialized/TeamSpeakClient.cs

```diff
@@ -620,6 +629,285 @@ public Task ChangeNickName(string nickName, WhoAmI whoAmI)
+        {
+            var res = await Client.Send("ftinitupload",
+                new Parameter("clientftfid", _fileTransferClient.GetFileTransferId()),
+                new Parameter("cid", channelId),
+                new Parameter("cpw", channelPassword),
+                new Parameter("name", NormalizePath(filePath)),
+                new Parameter("size", size),
+                new Parameter("overwrite", overwrite),
+                new Parameter("resume", 0)).ConfigureAwait(false);
+
+            var parsedRes = DataProxy.SerializeGeneric<InitUpload>(res).First();
+
+            await _fileTransferClient.SendFile(dataStream, size, parsedRes.Port, parsedRes.FileTransferKey).ConfigureAwait(false);
+
+            if (verify)
+            {
+                await VerifyUpload(parsedRes.ServerFileTransferId).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>Waits until the server fully receives the file or throws an exception when the upload times out.</summary>
+        private async Task VerifyUpload(int serverFileTransferId)
+        {
+            long arrivedBytes = 0;
+            var intervalMillis = 100;
+            var timeoutMillis = 3000;
+            var currentTimeoutMillis = intervalMillis * -1;
+
+            while (true)
+            {
+                var transfers = await GetCurrentFileTransfers();
+                var currentTransfer = transfers.Where(transfer => transfer.ServerFileTransferId == serverFileTransferId).FirstOrDefault();
+
+                if (currentTransfer == null)
+                {
+                    // Download finished
+                    return;
+                }
+
+                if (currentTransfer.SizeDone == arrivedBytes)
+                {
+                    // No upload progress
+                    currentTimeoutMillis += intervalMillis;
+
+                    if (currentTimeoutMillis > timeoutMillis)
+                    {
+                        try
+                        {
+                            await StopFileTransfer(serverFileTransferId).ConfigureAwait(false);
+                        }
+                        catch
+                        { }
+
+                        throw new FileTransferException("File upload timed out.");
+                    }
+                }
+                else
+                {
+                    // Upload progress
+                    currentTimeoutMillis = 0;
+                    arrivedBytes = currentTransfer.SizeDone;
+                }
+
+                await Task.Delay(intervalMillis).ConfigureAwait(false);
+            }
+        }
+
+        #endregion
+
+        #region DownloadFile
+
+        public Task<byte[]> DownloadFile(int channelId, string filePath) => DownloadFile(channelId, string.Empty, filePath);
+
+        public async Task<byte[]> DownloadFile(int channelId, string channelPassword, string filePath)
```

Comment:
Would it be possible to provide an API that returns a stream? Depending on the size of the file, it might be a bit heavy for the RAM.

Reply:
I thought about implementing a DownloadFile method that returns a stream and the size of the file. Correct me if I'm wrong but when returning the stream, the underlying TcpClient and socket would have to remain open and wouldn't be disposed by the caller of the method.

Reply:
I just looked at how the HttpClient of the .NET class library does it:
https://github.com/microsoft/referencesource/blob/master/System/net/System/Net/Http/HttpContent.cs

Basically, they just wrap it in a MemoryStream. Thats odd and I think that I missed something because that would mean that you cannot download files larger than 4GiB with the default HTTP client.

I'd consider returning a stream as future work. `byte[]` should be fine for now.